submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,293 @@
+
+In the current implementation, when a client Pod is scheduled, until it's annotated with a globalIP,
+and the corresponding egress rules are programmed on the node, it was not able to communicate with
+remote Services. This was creating some issues in-terms of user-experience. In this proposal, as we
+have a globalIP at the cluster level, a Pod can communicate with remote Services right away which
+will be closer to the Vanilla Submariner experience.
+
+## Design Details
+
+The following CRD is proposed to support the new design.
+
+```Go
+type GlobalnetEgressIP struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec is the specification of desired behavior of GlobalnetEgressIP object.
+    Spec               EgressIPSpec `json:"spec"`
+
+    // Observed status of GlobalnetEgressIP. Its a read-only field.
+    // +optional
+    Status             EgressIPStatus `json:"status,omitempty"`
+}
+
+type EgressIPSpec struct {
+    // GlobalIPs is a list of EgressIPs requested. The requested globalIPs should be part
+    // of globalCIDR of the Cluster. If the requested globalIPs are not allocated to other
+    // K8s objects, then it will be allocated and the status field will be updated accordingly.
+    // If the globalIPs are already allocated to other objects, creation of the CRD will fail.
+    // If this field is not specified, a single globalIP will be allocated.
+    // This field cannot be changed through updates.
+    // +optional
+    GlobalIPs  []string `json:"globalIPs,omitempty"`
+
+    // Selects the Namespaces to which this GlobalnetEgressIP object applies.
+    // If an empty namespaceSelector: {} is configured, it selects all the Pods in all namespaces.
+    // On the other hand, if you omit specifying namespaceSelector, it does not select any
+    // namespaces and selects only Pods from the namespace where the GlobalnetEgressIP is
+    // deployed to.
+    // +optional
+    NamespaceSelector metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+
+    // Selects the pods whose label match the definition. This is an optional field and
+    // in case it is not set, it results in all the Pods selected from the NamespaceSelector.
+    // In case it is set, its intersected with NamespaceSelector, and only Pods that match the
+    // PodSelector and present in the NamespaceSelector will have the GlobalIPs as EgressIPs.
+    // Either namespaceSelector or podSelector have to be specified. If both are omitted,
+    // its considered as an error.
+    // When a Pod or Namespace matches multiple GlobalnetEgressIP objects, there is no guarantee
+    // which of the globalIP address that are assigned to GlobalnetEgressIP will be used.
+    // +optional
+    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
+}
+
+type EgressIPStatus struct {
+    // The list of GlobalIPs assigned via this GlobalnetEgressIP object.
+    GlobalIPs []string `json:"globalIPs"`
```

Comment:
Same about the list

Reply:
Is this list expected to be different than the one in spec?

Reply:
Spec.GlobalIPs is an optional field. 

- If user explicitly specifies it and Globalnet is able to allocate those IPs, then this field will be same as Spec.GlobalIPs.

- If user explicitly specifies it but Globalnet **cannot** allocate those IPs, then this field will be empty and Status will be updated to Error (with the appropriate message).
- If user does not specify Spec.GlobalIPs, then this field includes the allocated globalIPs.

Reply:
Can't we have the controller fill that field if it's omitted?

Reply:
Well, we can theoretically have an annotation that marks if it was auto allocated. I don't think it matters much though, but maybe I'm wrong?

Reply:
Same as in my other question, what happens if we want multiple ips?

Reply:
The new GlobalnetIP + NumGlobalnetIps looks good to me.

Reply:
@mangelajo So shall we retain the allocated globalIPs (as a list) in the Status field (i.e., Status.GlobalIPs) or shall the controller just update Spec.GlobalIP and Spec.NumGlobalIPs itself (and remove Status.GlobalIPs)

Reply:
We can use Spec.GlobalIP for user allocated and Status.GlobalIP for actual allocated. So Spec contains what was configured and Status contains what is actually allocated.

Reply:
In the current CRD Status.GlobalIPs is a list. So if four globalIPs are allocated (assuming that user specified Spec.numGlobalIPs=4), then Status.GlobalIPs will be as follows

```
   status:
     globalIPs:
      - 169.254.0.100
      - 169.254.0.101
      - 169.254.0.102
      - 169.254.0.103
```

Do you think we can rather use something like this (which reduces the size of the object)
```
   status:
     globalIPs: "169.254.0.100-169.254.0.103"
```

Reply:
> If we're keeping `status` and the `GlobalIP` and `NumGlobalIPs` fields are optional then I'd expect them to be mirrored in the `status` field, instead of going with a different format.

I'm fine with this proposal if it's okay for others.

Reply:
Let's re-visit this after we have an agreement on removing the Spec.GlobalIP. If we remove Spec.GlobalIP, then we can reflect the allocated globalIPs in Status field as Status.GlobalIP and Status.NumGlobalIPs

Reply:
I would provide a list of IPs, otherwise it's harder to parse for the consumers, and if you have fragmentation you could end with things like:

169.254.0.1-169.254.0.3,169,169.254.0.9

or stuff like that.

I think a list is pretty straight forward, and not that consuming (let's avoid trying to optimize where is likely not necessary), I think usability is more important.

Reply:
Also, if we are going to remove the ability to request specific IPs, you need to talk to @mkimuram it was his request.

I think it's important-ish, because such IPs will be used as part of firewall rules which could be not automated to track this API. Of course requesting an specific IP can still fail because such IP is not available. But.. if we try to always use the lower-ips when selecting automatically, the admins could use the end of the range for manual selection.

Reply:
I'm afraid, we cannot support Fragmentation with iptables. The IPAddresses should be contiguous.

Reply:
One day (probably not that far) we will be doing only nftables. (we are already doing that in the background), or ovn rules...  may be we should not tailor the API to the implementation and make the API free of the details of the implementation as much as we can.

Reply:
In OVN for eaxmple it's more restrictive (CIDRs instead of IP ranges..)

Reply:
> One day (probably not that far) we will be doing only nftables. (we are already doing that in the background), or ovn rules... may be we should not tailor the API to the implementation and make the API free of the details of the implementation as much as we can.

But in that case the Spec needs to change no? Or is it enough just to ask for "X global ips" and the implementation will allocate what it can?

> In OVN for example it's more restrictive (CIDRs instead of IP ranges..)

That's interesting BUT is the globalnet processing tied to the CNI? I thought we're using iptables because it fits our needs, but not necessarily that it has to use the CNI.